atlasdb-feign/src/main/java/com/palantir/atlasdb/http/ExceptionCountingRefreshingClient.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2018 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.http;
+
+import java.io.IOException;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Supplier;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.palantir.logsafe.SafeArg;
+
+import feign.Client;
+import feign.Request;
+import feign.Response;
+
+public class ExceptionCountingRefreshingClient implements Client {
+    private static final Logger log = LoggerFactory.getLogger(ExceptionCountingRefreshingClient.class);
+
+    private static final long DEFAULT_EXCEPTION_COUNT_BEFORE_REFRESH = 10_000L;
+
+    private final Supplier<Client> refreshingSupplier;
+    private final long exceptionCountBeforeRefresh;
+    private final AtomicLong counter;
+
+    private volatile Client currentClient;
```

Comment:
Nice, I think you do need this (calls can go through multiple threads but this is only written in one, for a given refresh).